### PR TITLE
Introduce vertx-config module to enable configuration modularization and autoreload

### PR DESCRIPTION
**Pull request by marcinczeczko**
## Description
In order to improve how Knot.x configuration is to be managed a vertx-config module introduced together with the following features used:
- Auto-Redeploy knotx after configuration is changed
- Support for multiple formats of configuration file
- Favour the HOCON format of config files (human readable json) as the preferred one
- Ability to split config files between multiple files that can be glued together by includes of HOCON format

## Motivation and Context
- simplify the configuration file format (json has to much boilerplate)
- add ability to reload knotx on configuration change
- ability to chose a config format that works the best for your project (hocon, json, yaml)
- separate configuration files from the codebase

## Screenshots (if appropriate):
N/A

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the [code style](https://github.com/Cognifide/knotx/blob/master/CONTRIBUTING.md#coding-conventions) of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly. -> to be provided in #396

---
I hereby agree to the terms of the Knot.x Contributor License Agreement.

### CODE REVIEW

File: knotx-core/src/main/java/io/knotx/launcher/KnotxStarterVerticle.java [hunk 1]
```diff
@@ -16,56 +16,117 @@
...
+      if (!deployedModules.isEmpty()) {
+        LOGGER.warn("Configuration changed - Re-deploying Knot.x");
+        Observable.fromIterable(deployedModules)
+            .flatMap(item -> vertx.rxUndeploy(item.getDeploymentId()).toObservable())
+            .collect(() -> Lists.newArrayList(),
+                (collector, result) -> collector.add(result))
+            .subscribe(
+                success -> {
+                  LOGGER.warn("Knot.x STOPPED.");
+                  deployVerticles(conf.getNewConfiguration(), null);
```
Comment (Skejven) [inline knotx-core/src/main/java/io/knotx/launcher/KnotxStarterVerticle.java]: Won't there be NPE in `deployVerticles` in subscribe `onSuccess` / `onError`?

File: knotx-core/src/main/java/io/knotx/launcher/ModuleDescriptor.java [hunk 2]
```diff
@@ -96,4 +97,20 @@ public String toString() {
     }
     return result.toString();
   }
+
+  public enum DeploymentState {
+    UNKNOWN("Unknown state"),
+    SUCCESS("Deployed"), FAILED("Failed to deploy");
```
Comment (Skejven) [inline knotx-core/src/main/java/io/knotx/launcher/ModuleDescriptor.java]: New line here.

File: knotx-core/src/main/java/io/knotx/launcher/KnotxStarterVerticle.java (hunk 1 above)
Comment (marcinczeczko) [reply knotx-core/src/main/java/io/knotx/launcher/KnotxStarterVerticle.java]: good spot. fixed.

File: knotx-core/src/main/java/io/knotx/launcher/ModuleDescriptor.java (hunk 2 above)
Comment (marcinczeczko) [reply knotx-core/src/main/java/io/knotx/launcher/ModuleDescriptor.java]: fixed

File: knotx-example/knotx-example-app/config/application.conf [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+  adapter {
+    action = knotx.adapter.action.http
+    service = knotx.adapter.service.http
+  }
+}
+
+config {
+  server {
+    options.config.defaultFlow {
+      include file("./config/includes/server-default-flow.conf")
```
Comment (tomaszmichalak) [inline knotx-example/knotx-example-app/config/application.conf]: Let's imagine:
```
include file (data-provider.conf)
```
then in `data-provider.conf`
```
include file (data-provider-knot.conf)
include file (data-provider-adapter-http.conf)
```
Modularity looks very promising.

**Pull request merged**